### ISSUE TITLE
Filter out non-OD links

### ISSUE THREAD

**Issue by Chaphasilor** [Chaphasilor/odcrawler-scanner#5]
Maybe there's a certain pattern to the way scan results of non-ODs look (one dir, one file maybe?)

If such a result is detected, omit it from the reply (maybe mention that it was omitted)

**Comment by Krazybug** [comment]
Just a question, when it's not an OD, ODD should fail, no ?

**Comment by Chaphasilor** [comment]
nope: https://www.reddit.com/r/testingground4bots/comments/kew5zl/test/gg4u655?utm_source=share&utm_medium=web2x&context=3

**Comment by Krazybug** [comment]
Here is the algo of ODSHOT.

1. first avoid links in a blacklist
2. check that the link is working
3. go up step by step and check each of the parents is  reachable to get the working root.
4. Starting from the root, check all hrefs, open them and see if you can get the parent and check it contains one file which is not in these mimetypes ("text/html", "text/css", 'application/xml') .
5. for each of them, if you can't go up, call the functions recursively.
6. This way you get all the OD roots of a link (No guarantee for the domain) 


Some pathologic cases with JS in them remain but they are marginal. So I don't take site with JS into account. 

It you're not too much in a hurry, I can work on this issue as I proposed.

**Comment by Chaphasilor** [comment]
So you basically double-check if the root of each link is the actual root?

I imagine that takes quite some time for larger ODs 🤔 

And if I understand correctly, the algorithm also finds the most top-level OD on that domain, right? That's nice 👌🏻 

Take your time with it, we're not in a hurry. There are lots of things to work on, so for now I'm busy :)

**Comment by Krazybug** [comment]
Yes and no. Less than 2 hours to index the sub including the pushshift API with 250 green threads and my poor xDSL connection.
You almost always have less than 10 pages to explore. It's ODD in reverse mode ;-). It doesn't index them totally.

Yes the list of most top-levels ODs on a domain as sometimes the connection is refused on the domain names and more that one posts were reported in different subdirs. This way I can aggregate them.
But for your need the single one matching the link posted is enough.

**Comment by Krazybug** [comment]
I'm just understanding your question about performance. The hrefs I'm checking are just subdirs (html with relative path or absolute paths stripped with the current dir) not the other kind of links.

Yeah it's a bit tricky and it needed some tuning but it's reliable now.

**Comment by Chaphasilor** [comment]
I see. Sounds good!

When trying to implement that algorithm in JavaScript, you will have to use NodeJS though, because CORS doesn't allow you to make most requests from the browser :)

**Comment by Krazybug** [comment]
Thanks I guessed this was the case. Maybe I could also start with some unit or CLI tests with a local http server ?

**Comment by Chaphasilor** [comment]
You'd need a *lot* of test scenarios. Maybe using the live servers would be easier in this case.

**Comment by Krazybug** [comment]
You're right.  Let's start simple

**Comment by Chaphasilor** [comment]
hey @Krazybug any progress on this? :D

**Comment by Krazybug** [comment]
Hi @Chaphasilor,

Sorry I missed your notif. I'm a bit busy, professionally speaking, these days.
I'll try to work on it soon.

Would you accept a TS rather than vanilla JS implementation ?

**Comment by Chaphasilor** [comment]
Hey, no worries! Sure, TypeScript is fine as well :)

It could even be Pseudo-Code ¯\\\_(ツ)\_/¯
